matrix.go

```diff
@@ -8,42 +8,13 @@ import (
 	"net/url"
 )
 
+type Params map[string][]string
+type ParamKeys []string
+
 type Matrix struct {
 	Clients   []string
-	Servers   []string
-	Behaviors []string
-}
-
-type Results []Result
-
-func BeginMatrixTest(matrix Matrix) Results {
-
-	len := len(matrix.Clients) * len(matrix.Servers) * len(matrix.Behaviors)
-	results := make(Results, 0, len)
-
-	for _, client := range matrix.Clients {
-		for _, server := range matrix.Servers {
-			for _, behavior := range matrix.Behaviors {
-
-				testCase := TestCase{
-					Client:   client,
-					Server:   server,
-					Behavior: behavior,
-				}
-
-				result := ExecuteTestCase(testCase)
-				results = append(results, result)
-			}
-		}
-	}
-
-	return results
-}
-
-type TestCase struct {
-	Client   string
-	Server   string
-	Behavior string
+	ParamKeys ParamKeys
+	Params    Params
```

Comment:
Okay, so `ParamKeys` is a list of `SERVERS`, `BEHAVIORS`, etc., and `Params[k]` is the collection of values for key `k`. Sounds like these two pieces of information should be in one struct? Consider just creating a,

    type Dimension struct {
        Name string
        Values []string
    }

(I called it dimension because each of those adds a new dimension to the test matrix, but you can call it whatever.)